Add support and tests for new at-testsetup macro + usage

quinnj opened the pull request:
Relies on https://github.com/julia-vscode/TestItems.jl/pull/14 and https://github.com/julia-vscode/TestItemDetection.jl/pull/7.

Support not yet implemented in the vscode extension.

This is a quick prototype of some of the ideas presented in https://github.com/julia-vscode/TestItemRunner.jl/issues/11#issuecomment-1268721406. Namely, this (and above mentioned) PRs provide:
* A new `@testsetup Foo begin...` macro in TestItems.jl that allows wrapping a block of code in a module that multiple `@testitem`s can then share
* Support in TestItemDetections.jl for a new keyword argument to `@testitem`s like `setup=[Foo]` where `@testsetup` modules can be specified
* Support in TestItemRunner.jl for executing `@testitem`s with support for the new `setup=[Foo]` keyword arg by eval-ing a relative `using` to the previously defined `@testsetup` module.


A few quick observations since I kind of just jumped into the deep end here without worrying too much about broader details/context:
* The `@testsetup` macro is.....kind of worthless? As-is, it just expands to a module, but it's not really doing much that the user could just do themselves by writing `module Foo ...`
* BUT, it's also awkward just expanding to a regular module because now we're introducing an "order of operations" for `@testitem`s, whereas up till now they've always been completely runnable independently. So, for example, calling `@run_package_tests` fails if it's run _before_ any `@testsetup`s that are required by `@testitem`s, even though the `@testitem`s can be defined anywhere in any order. 

That's as far as I've gotten on thinking through things here; it does make me wonder if it would be possible/necessary to somehow include the `@testsetup`s in the static parsing as well? But it's not immediately clear to me how to do that without redefining the `@testsetup` module over and over again for each `@testitem` that depends on it. Or maybe we're ok introducing this bit of state since when you're using TestItemRunner.jl, you're thinking through structuring your tests like normal julia tests anyway (where order matters)?

davidanthoff commented:
> necessary to somehow include the @testsetups in the static parsing as well? 

Yep, I think that is the way to go.

> The @testsetup macro is.....kind of worthless?

With the detection of a `@testsetup` via parsing, the actual `@testsetup` macro would just return `nothing`, like the `@testitem` macro. The benefit would be that if a `@testsetup` is defined inline (i.e. in the actual package code) then none of that code runs when the package is loaded. So really the role of the `@testsetup` macro is to remove the code from normal package loading and make it detectable via static parsing.

> But it's not immediately clear to me how to do that without redefining the @testsetup module over and over again for each @testitem that depends on it.

This is meant to address the shared state story in #11, right? I think in that case what we would do is keep a list of testsetup modules that have been executed. Then once we start to execute test items, we check if it requires a setup, if it does, we check whether that particular testsetup has run or not. If it has, we just load that into the temporary module of the test item. If not, we actually execute the testsetup, add it to the list of things that have run and then inject into the temporary test module. Essentially the idea is that we only want these `@testsetup` blocks to run once, and then reuse that state for multiple `@testitem`s. Users will have to be careful to treat that state as read-only, really...

In the extension we will then have to also detect whenever there is a change to a `@testsetup` block, and if there is, we remove it from the dict of `@testsetup` code that has run, so that when we run another `@testitem` that needs it, it will rerun that `@testsetup`.

Does that make sense?

I should say, I actually like your approach with the `setup` keyword better than my original idea of a `@usingts Foo` because it makes it much easier to actually figure out what is really needed to execute a given `@testitem`.

nickrobinson251 commented:
API-wise, having to name a `@testsetup` is a bit annoying

If you have some `const`, that's a bit finickity (or expensive) to set-up and don't want to copy-paste it around, then with `Test` you'd just do
```
const HTTPBIN = get(ENV, "JULIA_TEST_HTTPBINGO_SERVER", "httpbingo.julialang.org")
```
then refer to `HTTPBIN` in whatever tests you want

But with `@testsetup` as it is, you know need to come up with another name like
```julia
@testsetup Stuff begin
    const HTTPBIN = get(ENV, "JULIA_TEST_HTTPBINGO_SERVER", "httpbingo.julialang.org")
end
```
or 
```julia
@testsetup HTTPBIN begin
    val = get(ENV, "JULIA_TEST_HTTPBINGO_SERVER", "httpbingo.julialang.org")
end
```

It's good that you know need to declare that a test is dependent on this shared const, but i wonder about a short hand for having single values being a "testsetup" e.g.
```
@testsetup HTTPBIN = get(ENV, "JULIA_TEST_HTTPBINGO_SERVER", "httpbingo.julialang.org")

@testitem "Content-Encoding" setup=[HTTPBIN] begin
    @test isfile(HTTP.download("https://$HTTPBIN/gzip"))
end

davidanthoff commented:
@nickrobinson251 I think your idea is somewhat similar to the global data storage idea in https://github.com/julia-vscode/TestItemRunner.jl/issues/11#issuecomment-1268721406? I think maybe we just need both, a simple global value store and a full blown test module story...

quinnj commented:
Ok, making progress here. I've changed the TestItems.jl `@testsetup` macro to return nothing, and updated the TestItemRunner.jl code to have a global, lockable `Set` we use to track evaled `@testsetup` modules. 

The remaining piece is to update TestItemDetection.jl to be able to statically parse `@testsetup` blocks.

NHDaly commented:
The other option to remove the need to name the `@testsetup` module, as @nickrobinson251 pointed out, would be to combine them into a scoping mechanism, rather than using named codeblocks.

This could also address #5, which I alluded to here: https://github.com/julia-vscode/TestItemRunner.jl/issues/11#issuecomment-1376052671

Something like this:
```julia
@testgroup "A name is optional here" begin

    @testsetup begin
        # This code is shared by all testitems in the group
    end

    @testitem "A" begin
        # ...
    end

    @testitem "B" begin
        # ...
    end

end
```

So then, the testitems are still independent, but they share this common nested structure just for naming purposes, and also they share common setup code.

I guess you could also do the shared globals in this way as well, and restrict them to a `@testgroup`?

----------------

What do you all think? It definitely does expand the testing taxonomy quite a bit, not least because there's also `@testset` and `@test` from Base.

nickrobinson251 commented:
> combine them into a scoping mechanism

I do quite like that idea -- it certainly looks neat in that example.

One possible downside i see is that usually tests are already organised some way -- most commonly into files testing related functionality -- and i think sharing set-up code is actually an orthogonal way of organising tests. i.e. Tests for very different functionality can still need the same set-up code. So i don't know if we'd want sharing-set-up-code to become the main determinant of how tests get grouped. Or at least, switching to such a grouping could be quite a lot of code churn. 

But i'd want to try switching over a few different packages tests to see if my suspicions are confirmed.

davidanthoff commented:
One reason I've been a bit skeptical about using scoping for anything with this new test stuff is that I think users would expect it to integrate with the existing module level scoping in Julia, and then things get _very_ complicated. So just as an example, once we start to use scoping for the test framework, I think users would expect this
```julia
module Foo

@testgroup Bar begin
   @testitem FooBar begin
    end
end

end
```
to essentially resolve to `Foo.Bar.FooBar`. But now what if there is another file that has
```julia
module FooledYou
    include("firstfile.jl")
end
```
I think my main intuition was that a) I don't want to integrate with the existing module level scoping, b) to have two scoping based hierarchy construction ways that exist in parallel but don't integrate is confusing.

Now, I might be wrong on that, but that was my thinking so far :) I think @nickrobinson251's suggestion of trying this out with more packages is the right way to try to resolve that.

quinnj commented:
> The other option to remove the need to name the `@testsetup` module, as @nickrobinson251 pointed out, would be to combine them into a scoping mechanism, rather than using named codeblocks.

I do think there are some merits to this approach, but the more I weighed the two options, the more I leaned towards the `@testsetup` approach of creating a reusable module. I like that the modules are more composable as independent blocks that can be reused across files/`@testitem`s with out forcing a hierarchy.

The other challenge I see with that approach is frankly the implementation: having to do static parsing via token stream makes it pretty challenging as-is to just parse a self-contained block. I'd worry about the robustness of detecting a top-level `@testgroup`, then parsing until.......we see a `@testitem`? What about if users try to put _additional_ code in between `@testitem`s within a `@testgroup`? Once we start putting things in a "scope", I think the downside is that we have to then start _respecting_ that scope and I worry people will expect code between the top `@testgroup` and the first `@testitem` would apply to that 1st `@testitem`, but then code in between the 1st and 2nd `@testitem` _wouldn't_ be viewable to the first `@testitem`? Just seems harder, IMO, to have clear semantics here, even ignoring the implementation.

davidanthoff commented:
> The other challenge I see with that approach is frankly the implementation

Yes, very much agreed!! The current detection logic is already fairly complicated (https://github.com/julia-vscode/TestItemDetection.jl/blob/acfec5294b2230d7b8d8359d77e4ce25d7a5a752/src/packagedef.jl#L6). Maybe there is a better way to write this kind of code, but it is painful, to put it mildly... Adding the `@testsetup` logic would be straightforward, I think one could essentially copy-paste that existing code and replace a few things, but having to deal with nesting etc. would significantly complicate things.

One other thought: we could start with the `@testsetup` story, implement that (well, @quinnj seems to be almost done in any case) and then take it from there? Adding a scoping stuff later is always possible, right?

NHDaly commented:
Awesome discussion! :) Thanks all, i'm pretty much convinced. Those are all great points, especially around the semantics. 👍 

Seems like the `@testsetup` approach would cover things well, and it's a good way to get started.

------

I do share @nickrobinson251's concern about the verbosity, but i guess it's not too bad. I do like the suggestion you had about the single-variable `@testsetup`. But i guess that could always be added later?

NHDaly commented:
So in this version, can the `@testsetup` macro be anywhere, in any file, and then it can get used from any other file? What are the boundaries?

----

Also: How is this going to work with our subpackages in the same repository? Can you share `@testsetup`s across subpackages? Would we want that?

Also also: what happens if you have multiple `@testsetup`s with the same name? Do they union together into one big setup? Or throw an error?  I think an error makes sense...? 🤔 😬  A lot to think about!

quinnj commented:
Ok, the tests are now passing for me locally moving to statically parsing the `@testsetup` blocks. @davidanthoff, can you help me understand the vendored setup here? I updated the PR to TestItemDetection.jl w/ the new `@testsetup` code, but it's not clear whether I should copy-paste the code to the file in this repo or if you have a separate git submodule situation going on? If you can give me some pointers, I can update the PR here to get CI happy.

> So in this version, can the @testsetup macro be anywhere, in any file, and then it can get used from any other file? What are the boundaries?

Yes, basically we have the same freedom as `@testitem` in that the `@testsetup` macro just returns `nothing`, and then TestItemDetection.jl does the static parsing of the blocks for TestItemRunner.jl. And because we statically parse all files in a directory, order doesn't matter either; we statically detect all the `@testitem` and `@testsetup` blocks, then execute the `@testitem`s and eval the `@testsetup`s as needed, which has the nice property that if a `@testsetup` is never used by a `@testitem`, it won't be eval-ed!

> Also: How is this going to work with our subpackages in the same repository? Can you share @testsetups across subpackages? Would we want that?

This is a good question and worth thinking about. Currently the `@run_package_tests` macro only looks at the immediate Project directory, so it wouldn't share anything outside the project. From a parent project, however, it would indeed pick up any subproject `@testsetup` and `@testitem`s currently. The filtering ability means we could easily restrict things to only eval `@testitem`s of the immediate project, but it also seems handy to be able to run child tests w/ a parent. 

> Also also: what happens if you have multiple @testsetups with the same name? Do they union together into one big setup? Or throw an error? I think an error makes sense...? 🤔 😬 A lot to think about!

Also a great question. In the current PR, I believe it would just replace an existing `@testsetup` w/ the same name, which doesn't seem ideal. I agree that throwing an error is probably best, though I somewhat worry about the subproject case you mentioned where suddenly a child or parent tests start failing because a parent/child happened to define a `@testsetup` w/ the same name.

davidanthoff commented:
> or if you have a separate git submodule situation going on? If you can give me some pointers, I can update the PR here to get CI happy.

Yep, git submodules is what we use here. I just gave you write access to the TestItemDetection.jl repo. I would push your branch for TestItemDetection.jl to that origin repo (rather than your fork), then in this repo here on the branch for this PR, go to `packages/TestItemDetection`, run `git fetch`, and then `git checkout YOURBRANCHNAME`. I can't think of a way to use the branch on your fork right now, so I think you'll just have to get that into the origin repo for TestItemDetection first.

> Also: How is this going to work with our subpackages in the same repository? Can you share @testsetups across subpackages? Would we want that?

My gut feeling is probably not? The entire env setup part really only takes into account the immediate parent package definition, so we should probably scope it to that? 

> > Also also: what happens if you have multiple @testsetups with the same name? Do they union together into one big setup? Or throw an error? I think an error makes sense...? 🤔 😬 A lot to think about!
>
> Also a great question. In the current PR, I believe it would just replace an existing @testsetup w/ the same name, which doesn't seem ideal. I agree that throwing an error is probably best, though I somewhat worry about the subproject case you mentioned where suddenly a child or parent tests start failing because a parent/child happened to define a @testsetup w/ the same name.

I think just completely segmenting by package would solve that, right?

NHDaly commented:
I agree with you both: my preference would be to completely segment by package.

But how do you enforce that? My impression is that right now TestItemsRunner / VSCode search _anywhere in the current directory_ for TestItems, not just in `src/` and `test/`.

Our monorepo is set up as one "parent project," with a bunch of subpackages in `packages/`, like this:
```
.
├── Project.toml
├── Manifest.toml
├── src/
│   └── TopLevelPackage.jl
├── test/
│   └── runtests.jl
└── packages/
    ├── SubPackageA/
    │   ├── src/
    │   │   └── SubPackageA.jl
    │   ├── test/
    │   │   └── runtests.jl
    └── SubPackageB/
        ├── src/
        │   └── SubPackageB.jl
        └── test/
            └── SubPackageB.jl
```
So if we had a `@testsetup` in `./packages/SubPackageB/src/SubPackageB.jl`, would it get picked up by a `@testitem` in `./test/runtests.jl`? I wouldn't want it to.

Does that make sense?

davidanthoff commented:
The test runner in the extension at the moment detects the test item _and_ the project/package it should be run in. So in your example (assuming the sub packages each have a `Project.toml`), it would actually launch a separate test runner process for the tests in sub package A and sub package B, because they would have to run in a different environment. So while the UI surfaces all test items, when they are run in VS Code, they are actually already completely segmented by package, in the sense of different test processes. The rule is simple: for each `@testitem` we look in the folder, then in the parent folder etc until we find a `Project.toml` that defines a package, and then the `@testitem` is part of that package.

So I think for the `@testsetup` we could use exactly the same logic: we detect to which package each `@testsetup` belongs, and then it can only be used from `@testitem`s that belong to the same package.

I am realizing, though, that the TestItemRunner.jl implementation doesn't follow that philosophy fully, I think what is missing there is that we should by default skip any child folder that has a new `Project.toml` in it that defines a new package. So in @NHDaly's tree, if someone runs `test/runtests.jl`, then that should actually not run the `@testitem`s of the sub packages, right?

quinnj commented:
> I am realizing, though, that the TestItemRunner.jl implementation doesn't follow that philosophy fully, I think what is missing there is that we should by default skip any child folder that has a new `Project.toml` in it that defines a new package. So in @NHDaly's tree, if someone runs `test/runtests.jl`, then that should actually not run the `@testitem`s of the sub packages, right?

Yeah, that sounds right to me.

davidanthoff commented:
@quinnj Actually, I just realized that this is _not_ using git submodules, but git subtrees... I'm going to push a commit to this branch here that just picks up the new TestItemDetection stuff, and then you can take over again, ok?

davidanthoff commented:
I just updated the extension and the LS to work with the new TestItemDetection version that also detects `@testsetup`. At the moment the extension doesn't do anything with that info, so I still need to incorporate it into the runner there, but I think the detection part on the extension side should be done.

davidanthoff reviewed (commented):
This is surprisingly not very involved at all, nice :) I left two comments, both hopefully fairly easy.

NHDaly commented:
@davidanthoff: +1, your explanation sounds perfect. Thanks for clarifying it! :)

Also: Woops, you're right, I had forgotten the `Project.toml`s in the subpackages. I've added them in now. Thanks! :)

quinnj commented on src/TestItemRunner.jl:
```diff
@@ -59,7 +59,29 @@ end
+    lock::ReentrantLock
+end
+
+const TEST_SETUP_MODULE_SET = TestSetupModuleSet(Set{Symbol}(), ReentrantLock())
```
Hmmmm, while I agree with the sentiment, I'm not sure it makes a difference in practice, since we're eval-ing the modules into `Main.TestSetups` anyway? or maybe we create a new run_tests module that we eval test setups into each run. I'll try that.

quinnj commented:
Ok, pushed an update; thanks @davidanthoff for doing the TestItemDetection update for me and combining the two syntax passes.

In the latest commit, we will now eval a unique "test setup module" per `run_tests` call, and any `@testsetup` blocks will be eval-ed into a module that lives in that per-`run_tests` module. The `@testitem` modules will then have the appropriate relative `using` inserted if they have `@testsetup` dependencies.

I like this solution because it avoids having global modules and makes independent `run_tests` invocations run in isolation with respect to test setups.

quinnj commented:
Ah yes, we'll need https://github.com/julia-vscode/TestItems.jl/pull/14 1st + tagged before the tests will pass here.

quinnj commented:
On a related note, @davidanthoff, do we really need the subtree setup in this package? I understand the motivation for TestItem.jl and TestItemDetection.jl, but it seems like TestItemRunner.jl doesn't share the same concerns and can be treated more like normal package. It would make development on this package simpler and running tests faster (since we can incrementally compile the CSTParser, TestItemDetection, etc. pieces and only recompile TestItemRunner.jl itself)

nickrobinson251 commented:
this is looking great!

I took it out for a spin (on HTTP.jl tests) and at first i found the `@testsetup` a bit clunky, because (similar to my comment above) I had to write the testsetup like
```diff
- const foo = "..."
+ @testsetup SHARED begin
+     const foo = "..."
+ end
```
_but then also_ change the code _within_ my testitem to be like
```diff
 @testitem "foo tests" setup=[SHARED] begin
-    @test HTTP.get("http://$foo") == expected
+    @test HTTP.get("http://$(SHARED.foo)") == expected
```
which felt ugly

BUT then i remembered "`@testsetup` is just a module", and realised i could `export foo` like
```diff
 @testsetup SHARED begin
+    export foo   
     const foo = "..."
 end
```
and now my code _inside_ the testitem can stay as before i.e.
```
 @testitem "foo tests" setup=[SHARED] begin
    @test HTTP.get("http://$foo") == expected
 end
 ```
 
 All of this made me think: **Why don't we just make it really clear that `@testsetup` creates a module?** That way, it'll be _obvious_ that you can `export` things.
 
How about we use this spelling:

```julia
@testsetup module SetupName 

end 
```
and then
```julia
@testitem "foo" using=[SetupName] begin

end
```
i.e. require `@testsetup` be followed by `module` (since we are just declaring a module), and make the `@testitem` keyword `using` (since we are just `using $mod`)

davidanthoff commented:
> BUT then i remembered "`@testsetup` is just a module", and realised i could export `foo`

Ah, that is cool, I hadn't thought about that aspect at all!

> How about we use this spelling:
>
> ```
> @testsetup module SetupName 
>
> end 

Or we could just call the macro `@testmodule`? That was the initial idea in #11.

> ```
> @testitem "foo" using=[SetupName] begin
>
> end

I like that!

One other thing I've been contemplating is whether we also need a tear-down option? It could be an optional second code block, say
```julia
@testmodule Foo begin
  # This sets things up
end begin
  # Here one can put cleanup code
end
```
But the semantics of when that tear-down code should run are a bit tricky in the VS Code extension...

nickrobinson251 commented:
> Or we could just call the macro @testmodule?

A possible issue with "`@testmodule`" as the name is that it implies "this is the module where your tests go" (at least, that's how it reads to me, and I think @NHDaly has mentioned the same concern), whereas having "setup" in the name clarifies better what this module is

davidanthoff commented:
> do we really need the subtree setup in this package?

Yeah, I think so, unfortunately. Otherwise, we force the test runner and the package to be tested to use the same version of the packages in the `packages` folder, right? That seems like the kind of package compat problem that we've tried to avoid at all costs in the VS Code extension.

davidanthoff commented:
> A possible issue with "@testmodule" as the name is that it implies "this is the module where your tests go" (at least, that's how it reads to me, and I think @NHDaly has mentioned the same concern), whereas having "setup" in the name clarifies better what this module is

Ah, good point. Also, thinking about a teardown option, in that case it might also be better to stick with `@testsetup`.

davidanthoff commented:
So on the call today we all gravitated towards @nickrobinson251's idea of making it
```julia
@testsetup module Foo
end
```
I think that requires some changes in TestItemDetection.jl only.

I'm about to tag a new TestItems with the new macro in there.

One other thing: I think we all also agreed that we would do parallelism via separate processes, in which case we don't need the lock, right?

NHDaly commented:
FYI: Jacob is out sick now, probably with strep. @nickrobinson251 are you able to push to Jacob's fork by any chance?

NHDaly commented:
If not, we can maybe move on to the next piece of work while we wait for this one?

nickrobinson251 commented:
i can't push to Jacob's fork, but i can PR it from my fork 😂 
https://github.com/quinnj/TestItemRunner.jl/pull/1

quinnj commented:
CI looks good here now; ready to merge!

davidanthoff reviewed (approved):
Very cool :)

KeithWM commented:
Given the difficulty for both `dev --local` and `add ./test/TestItemRunner` in one's `test` environment, it would be really nice if this improvement could be released and registered soon. What is the plan for v0.2.2? I really want to try TestItemRunner for the repo's in our industrial Julia application, but I do have some packages that require some shared setup between tests.

davidanthoff commented:
I think I want to hold off with releasing this for good until we have it fully implemented across all the clients where we need it, so essentially I'd like to register/release when this works here, in the Julia extension for VS Code and in the new https://github.com/julia-vscode/TestItemRunner2.jl. Primarily so that we make sure it all works in all clients, otherwise we could end up in a situation where we release things here, then release that there is some problem with the design in the extension but then we are restricted by backwards compatibility to address that.

Having said that in https://github.com/julia-vscode/TestItemRunner2.jl it should all work already, so that might be the easiest to try things out right now.

KeithWM commented:
I don't think TestItemRunner2 has been registered. I don't think there's a way to add unregistered packages to the test environment (and have the environment merge work), is there?
